libsolidity/ExpressionCompiler.cpp

```diff
@@ -811,36 +812,50 @@ bool ExpressionCompiler::visit(IndexAccess const& _indexAccess)
 			}
 		else
 		{
-			// stack: <base_ref> [storage_byte_offset] <index>
-			if (location == ArrayType::Location::Storage)
-				//@todo use byte offset, remove it for now
-				m_context << eth::Instruction::SWAP1 << eth::Instruction::POP;
-			u256 elementSize =
-				location == ArrayType::Location::Storage ?
-					arrayType.getBaseType()->getStorageSize() :
-					arrayType.getBaseType()->getCalldataEncodedSize();
-			solAssert(elementSize != 0, "Invalid element size.");
-			if (elementSize > 1)
-				m_context << elementSize << eth::Instruction::MUL;
+			// stack: <base_ref> <index>
+			m_context << eth::Instruction::SWAP1;
 			if (arrayType.isDynamicallySized())
 			{
 				if (location == ArrayType::Location::Storage)
-				{
-					m_context << eth::Instruction::SWAP1;
 					CompilerUtils(m_context).computeHashStatic();
-				}
 				else if (location == ArrayType::Location::Memory)
 					m_context << u256(32) << eth::Instruction::ADD;
 			}
-			m_context << eth::Instruction::ADD;
+			// stack: <index> <data_ref>
 			switch (location)
 			{
 			case ArrayType::Location::CallData:
+				m_context
+					<< eth::Instruction::SWAP1 << arrayType.getBaseType()->getCalldataEncodedSize()
+					<< eth::Instruction::MUL << eth::Instruction::ADD;
 				if (arrayType.getBaseType()->isValueType())
 					CompilerUtils(m_context).loadFromMemoryDynamic(*arrayType.getBaseType(), true, true, false);
 				break;
 			case ArrayType::Location::Storage:
-				m_context << u256(0); // @todo
+				m_context << eth::Instruction::SWAP1;
+				if (arrayType.getBaseType()->getStorageBytes() <= 16)
+				{
+					// stack: <data_ref> <index>
+					// goal:
+					// <ref> <byte_number> = <base_ref + index / itemsPerSlot> <(index % itemsPerSlot) * byteSize>
+					unsigned byteSize = arrayType.getBaseType()->getStorageBytes();
+					unsigned itemsPerSlot = 32 / byteSize;
+					m_context << u256(itemsPerSlot) << eth::Instruction::SWAP2;
```

Comment:
Yes, such arrays are absolutely nonsensical and forbidden at the type level.
An assert would not hurt, though.